experimental/components/src/units/converter_factory.rs

```diff
@@ -0,0 +1,339 @@
+            None
+        }
+    }
+
+    fn apply_si_prefix(si_prefix: &SiPrefix, ratio: &mut Ratio<BigInt>) {
+        match si_prefix.base {
+            Base::Decimal => {
+                *ratio *= Ratio::<BigInt>::from_integer(10.into()).pow(si_prefix.power as i32);
+            }
+            Base::Binary => {
+                *ratio *= Ratio::<BigInt>::from_integer(2.into()).pow(si_prefix.power as i32);
+            }
+        }
+    }
+
+    fn compute_conversion_term(
+        &self,
+        unit_item: &MeasureUnitItem,
+        sign: i8,
+    ) -> Option<Ratio<BigInt>> {
+        let conversion_info = self.payload.convert_infos.get(unit_item.unit_id as usize);
+        debug_assert!(conversion_info.is_some(), "Failed to get conversion info");
+        let conversion_info = conversion_info?;
+
+        let mut conversion_info_factor = Self::extract_ratio_from_unaligned(
+            &conversion_info.factor_sign,
+            conversion_info.factor_num(),
+            conversion_info.factor_den(),
+        );
+
+        Self::apply_si_prefix(&unit_item.si_prefix, &mut conversion_info_factor);
+        conversion_info_factor = conversion_info_factor.pow((unit_item.power * sign) as i32);
+        Some(conversion_info_factor)
+    }
+
+    fn extract_ratio_from_unaligned(
+        sign_ule: &SignULE,
+        num_ule: &ZeroSlice<u8>,
+        den_ule: &ZeroSlice<u8>,
+    ) -> Ratio<BigInt> {
+        let sign = Sign::from_unaligned(*sign_ule).into();
+        Ratio::<BigInt>::new(
+            BigInt::from_bytes_le(sign, num_ule.as_ule_slice()),
+            BigInt::from_bytes_le(num_bigint::Sign::Plus, den_ule.as_ule_slice()),
+        )
+    }
+
+    /// Creates a converter for converting between two single or compound units.
+    /// For example:
+    ///    1 - `meter` to `foot` --> Simple
+    ///    2 - `kilometer-per-hour` to `mile-per-hour` --> Compound
+    ///    3 - `mile-per-gallon` to `liter-per-100-kilometer` --> Reciprocal
+    ///    4 - `celsius` to `fahrenheit` --> Needs an offset
+    ///
+    /// NOTE:
+    ///    This converter does not support conversions between mixed units,
+    ///    such as, from "meter" to "foot-and-inch".
+    pub fn converter(
+        &self,
+        input_unit: &MeasureUnit,
+        output_unit: &MeasureUnit,
+    ) -> Option<UnitsConverter> {
+        let is_reciprocal = self.is_reciprocal(input_unit, output_unit)?;
+
+        // Determine the sign of the powers of the units from root to unit2.
+        let root_to_unit2_direction_sign = if is_reciprocal { 1 } else { -1 };
+
+        let mut conversion_rate = Ratio::one();
+        for input_item in input_unit.contained_units.iter() {
+            conversion_rate *= Self::compute_conversion_term(self, input_item, 1)?;
+        }
+
+        for output_item in output_unit.contained_units.iter() {
+            conversion_rate *=
+                Self::compute_conversion_term(self, output_item, root_to_unit2_direction_sign)?;
+        }
```

Comment:
Question: can `is_reciprocal` and `conversion_term` be computed at the same time to reduce work, or are they sufficiently separable? I quite like this new code, but it makes it look like `is_reciprocal` is significantly more complex of an operation than `conversion_term`.

Reply:
Computing `is_reciprocal` and `conversion_terms` are two separate operations and depend on each other.
In `is_reciprocal`, we compare all the powers in the base units of the input unit and the output unit. Additionally, the reciprocality is used in computing the conversion rates from the input unit to the output.

Given that we are using a zero-copy strategy, accessing the information will cost nothing. Splitting these functionalities is going to make the code more maintainable and readable.